File: samples/compose/app/src/main/kotlin/dev/getmilano/sample/designsystem/StyledText.kt
package dev.getmilano.sample.designsystem

import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.LiveRegionMode
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.liveRegion
import androidx.compose.ui.semantics.semantics

/**
 * Semantic text: the role states intent, this design system decides looks
 * and the matching assistive-technology semantics (a title is a heading;
 * a live region announces its changes).
 */
data class TextModel(
    val text: String,
    val role: Role = Role.BODY,
    val liveRegion: LiveRegion? = null,
) {
    enum class Role { TITLE, SUBTITLE, BODY }

    enum class LiveRegion { POLITE, ASSERTIVE }
}

@Composable
fun StyledText(model: TextModel) {
    val modifier =
        Modifier.semantics {
            if (model.role == TextModel.Role.TITLE) heading()
            when (model.liveRegion) {
                TextModel.LiveRegion.POLITE -> {
                    liveRegion = LiveRegionMode.Polite
                }

                TextModel.LiveRegion.ASSERTIVE -> {
                    liveRegion = LiveRegionMode.Assertive
                }

                null -> {}
            }
        }
    when (model.role) {
        TextModel.Role.TITLE -> Text(model.text, style = MaterialTheme.typography.titleLarge, modifier = modifier)
        TextModel.Role.SUBTITLE -> Text(model.text, style = MaterialTheme.typography.bodyMedium, modifier = modifier)
        TextModel.Role.BODY -> Text(model.text, style = MaterialTheme.typography.bodyLarge, modifier = modifier)
    }
}

File: samples/swiftui/Sources/DesignSystem/StyledText.swift
import SwiftUI

#if canImport(UIKit)
    import UIKit
#endif

/// Semantic text: the role states intent, this design system decides looks
/// and the matching assistive-technology semantics (a title is a heading;
/// a live region announces its changes).
struct TextModel {
    enum Role {
        case title
        case subtitle
        case body
    }

    enum LiveRegion {
        case polite
        case assertive
    }

    var text: String
    var role: Role = .body
    var liveRegion: LiveRegion?
}

struct StyledText: View {
    let model: TextModel

    var body: some View {
        styled
            .accessibilityAddTraits(model.role == .title ? .isHeader : [])
            .onChange(of: model.text) { newValue in
                // SwiftUI has no live regions; the closest honest mapping
                // is announcing the new value when it changes.
                guard model.liveRegion != nil, !newValue.isEmpty else { return }
                #if canImport(UIKit)
                    UIAccessibility.post(notification: .announcement, argument: newValue)
                #endif
            }
    }

    @ViewBuilder
    private var styled: some View {
        switch model.role {
        case .title:
            Text(model.text).font(.title3.weight(.semibold))
        case .subtitle:
            Text(model.text).font(.subheadline).opacity(0.85)
        case .body:
            Text(model.text).font(.body)
        }
    }
}
